### ISSUE TITLE
matrix equality check is incorrect

### ISSUE THREAD

**Issue by mikestillman** [Macaulay2/M2#938]
Matrix == Matrix currently checks that the sources and targets are the same module, which ignores the bases of the source or target, e.g. in the case when they are image modules.

Here are tests which currently fail:
```
  debug Core
  R = ZZ/101[a..d]
  M = image vars R
  M1 = image matrix{{b,a,c,d}}
  M == M1
  N = R^{4:-1}
  f0 = map(R^1,, matrix{{a,b,c,d}})
  -- example 1: sources are identical, targets are same image module.
  f = inducedMap(M,source f0,f0)
  g = inducedMap(M1,source f0,f0)
  assert(source f === source g)
  assert(source f == source g)
  assert(raw super f === raw super g)
  assert(f == g)
  -- example 2: sources are identical, targets are same subquotient
  --  (but with different image parts).
  M = (image vars R)/(image matrix{{c,d}})
  M1 = (image matrix{{b,a,c,d}})/(image matrix{{d,c}})
  assert(M == M1)
  assert(M =!= M1)
  f = inducedMap(M,source f0,f0)
  g = inducedMap(M1,source f0,f0)
  assert(source f === source g)
  assert(source f == source g)
  assert(raw super f === raw super g)
  assert(f == g)
  -- next case: sources are the same image, but with different gen order
  M = image vars R  
  N = image matrix {{a,b,c,d}}
  N1 = image matrix {{c,a,b,d}}
  f = map(R^1, N, {{a,b,c,d}})
  g = map(R^1, N1, {{c,a,b,d}})
  assert isWellDefined f
  assert isWellDefined g
  assert(f == g)
```

Here is a proposed bug fix:
```
Matrix == Matrix := (f,g) -> (
    -- this next test also covers the case when the source 
    -- and target are free.
    sameSources := source f === source g;
    if target f === target g and sameSources then 
        return (raw f === raw g);
    if target f != target g or source f != source g then
        return false;
    if sameSources then
        return(raw super f === raw super g);
    raw(super f * inducedMap(source f, source g)) === raw super g
    )
```

@DanGrayson , could you review this change?

@ggsmith and I found this bug, and created this proposed fix.

**Comment by DanGrayson** [comment]
Re: "Matrix == Matrix currently checks that the sources and targets are the same module"

Why is that a bad thing, mathematically?

**Comment by mikestillman** [comment]
That is not a bad thing, it is a good thing.  The problem is that then it checks that raw m === raw n, which is not correct (one can have the same map, without the raw matrices being identical, as in the examples we gave).

So my wording was not the best, the problem is that some maps which are the same mathematically come out not to be ==.

**Comment by DanGrayson** [comment]
Why isn't it a bug that `raw m === raw n` is false?

**Comment by mikestillman** [comment]
The raw matrix is simply a matrix between free modules in the engine (corresponding to the covers of the modules).  The problem is more that two matrices can be equal even when this matrix between covers is not equal.  A simple example is the first one I gave: there are two image modules M, M1.  These are the same module, but have different generators (permutations of each other).

**Comment by DanGrayson** [comment]
Ah, yes, the point I wasn't appreciating is that in this code
```
Matrix == Matrix := (m,n) -> target m == target n and source m == source n and raw m === raw n
```
the source and target are tested with == instead of ===.

**Comment by DanGrayson** [comment]
How did we live with this all these years?

**Comment by mikestillman** [comment]
I’m very surprised about that too!

**Comment by DanGrayson** [comment]
Your code looks clearer without `return` and negation:
```
Matrix == Matrix := (f,g) -> (
     if source f === source g
     then if target f === target g 
	  then raw f === raw g
	  else target f == target g and
	       raw super f === raw super g
     else source f == source g and
 	  target f == target g and 
	  raw(super f * inducedMap(source f, source g)) === raw super g)
```

**Comment by mikestillman** [comment]
Yes, that does look better!

I wonder if our changing the code will break any examples or tests?  I hope not!

**Comment by DanGrayson** [comment]
Well, there are two ways to find out: check it in and wait for the overnight run, or rebuild locally.

I think it will be fine.  In fact, I bet if we changed it to this:
```
Matrix == Matrix := (f,g) -> f === g
```
that no test would fail.

**Comment by DanGrayson** [comment]
Why don't you add those tests to a new file and put my version of your code in and we'll wait overnight?

**Comment by mikestillman** [comment]
OK, I’ll do that.

**Comment by mikestillman** [comment]
This result though would be incorrect still!  (Still, probably no test would fail there either).

**Comment by DanGrayson** [comment]
Probably the reason this code is never used is that math is done the way category theory is done: first you make two objects M and N of a category, and then you construct arrows from M to N.  It never really happens that you have to *prove* two objects are equal, and then make use of that to prove two arrows are equal.

**Comment by mikestillman** [comment]
The reason it came up for us: we made a function for checking that a pair of maps of complexes formed a short exact sequence.  But checking simply that ker alpha == image beta was failing, and we tracked it down to this.

**Comment by DanGrayson** [comment]
Hmm?  But that is a test that two modules are equal.

**Comment by mikestillman** [comment]
Those are complexes...

**Comment by DanGrayson** [comment]
Yes, but those are two subcomplexes of the same complex, so checking equality amounts to checking equality of a bunch of pairs of submodules.  The maps don't have to be checked.  Or were you checking them anyway?

**Comment by DanGrayson** [comment]
Were you also wasting time computing the differentials on ker alpha and image beta?

**Comment by mikestillman** [comment]
Yes.  But it was really just a check that our code was correct.

**Comment by DanGrayson** [comment]
Okay.

I guess that gives a good example where my comment about category theory is wrong!  :-)

**Comment by mikestillman** [comment]
Fixed in commit d232884b374f

**Comment by DanGrayson** [comment]
fixed by Mike in commit d232884b37
